internal/transform/scope-css.go

```diff
@@ -30,229 +30,12 @@ outer:
-							isKeyframes = false
-						}
-						out += strVal
-					case "*":
-						if parenCount == 0 && !isGlobal {
-							out += scopeRule("", opts) // turns "*" into ".astro-XXXXXX" rather than "*.astro-XXXXXX"
-						} else {
-							out += strVal
-						}
-					default:
-						// handle IDs with parens attached
-						if strings.Contains(strVal, "(") {
-							parenCount++ // if new paren opened, count it
-							isElement = true
-							isPseudoState = false
-						}
-
-						// if this is an element, check if it’s <body>, etc.
-						if isElement && globalElement(strVal) {
-							isGlobalElement = true
-						}
-
-						// whitespace tokens are used to reset chained classes and functions
-						if val.TokenType == css.WhitespaceToken {
-							// important: global elements like <body> may have classes that should not be scoped
-							if isElement && isGlobalElement {
-								isGlobalElement = false
-							}
-
-							// important: :global() might be chained (:global().some-class)
-							// so keep it active until whitespace is reached after final paren
-							if isGlobal && parenCount == 0 {
-								isGlobal = false
-							}
-						}
-
-						// scope class
-						isCssSelector := (gt == css.BeginRulesetGrammar || gt == css.QualifiedRuleGrammar) && (val.TokenType == css.IdentToken || val.TokenType == css.HashToken)
-						if isCssSelector && // don’t scope @media and other non-class specifiers
-							!isPseudoState && // don’t scope pseudostates
-							!isGlobal && // don’t scope in :global() scope
-							!isGlobalElement &&
-							!isBracket && // don’t scope within element brackets
-							parenCount == 0 { // don’t scope within parens like :not()
-							out += scopeRule(strVal, opts)
-						} else {
-							// otherwise, append output
-							out += strVal
-						}
-
-						// reset state
-						isElement = true
-						isPseudoState = false
-						declaration = ""
-					}
-				}
-
-				// generate tail of next rule
-				switch gt {
-				case css.BeginAtRuleGrammar,
-					css.BeginRulesetGrammar:
-					out += "{"
-				case css.DeclarationGrammar,
-					css.EndRulesetGrammar,
-					css.EndAtRuleGrammar:
-					out += ";"
-				case css.QualifiedRuleGrammar:
-					out += ","
-				}
-			default:
-				strData := string(data)
-				out += strData
-				for _, val := range p.Values() {
-					strVal := string(val.Data)
-					// handle CSS variables
-					if strings.HasPrefix(strData, "--") {
-						out += ":"
-					}
-					out += strVal
-				}
-				out += ";"
-			}
-		}
-		n.FirstChild.Data = out
+		// Use vendored version of esbuild internals to parse AST
+		tree := css_parser.Parse(logger.Log{AddMsg: func(msg logger.Msg) {}}, logger.Source{Contents: n.FirstChild.Data}, css_parser.Options{MinifySyntax: false, MinifyWhitespace: true})
+		// esbuild's internal `css_printer` has been modified to emit Astro scoped styles
+		result := css_printer.Print(tree, css_printer.Options{MinifyWhitespace: true, Scope: opts.Scope})
```

Comment:
This is pretty much it. Rather than modify the `tree` AST (too complex), we modified the `css_printer` to handle our scoping when printing selector nodes.